### ISSUE TITLE
Feature Request: Prefetch into Cache up to a Configurable Size

### ISSUE THREAD

**Issue by gordan-bobic** [rclone/rclone#995]
It would be very useful to have a configurable prefetch feature that buffers down into page cache, e.g. --max-read-ahead of data from the file on every read.

This would be particularly useful for cases where the upstream cloud storage provider seems to be throttling based on the number of requests/hour rather than MB/hour, (e.g. ACD). If the subsequent read request can be served from the cache, no further read happens. If a back-end read is required, prefetch an extra max-read-ahead sized block. By serving several consecutive requests from the prefetch cache it would greatly reduce the number of requests for use-cases such as media servers that only pass a small amount of data at a time up to the clients.

**Comment by ncw** [comment]
Provided the reader isn't seeking, rclone will read the files using streaming, so that will be just one request.

There is a certain amount of read-ahead - I did try adding more but it then can interfere with the seeking.

I need to revisit this!

**Comment by gordan-bobic** [comment]
The access pattern exhibited by media streaming server such as plex is such that they open a file, read a chunk at the required position and close the file. The read the next chunk and close the file, repeat. AFAICT, this is largely due to media clients behaving in that same way (they'll request a chunk of an certain length from a certain position and drop the TCP connection when they have had their fill). This is where prefetch would be very handy to avoid some number of the subsequent reads.

**Comment by ncw** [comment]
> The access pattern exhibited by media streaming server such as plex is such that they open a file, read a chunk at the required position and close the file. The read the next chunk and close the file, repeat. AFAICT

Ok that is interesting.  Have you verified that by looking at the rclone mount logs?

**Comment by gordan-bobic** [comment]
I'll try to put some rclone and inotify logs together to verify it for the specific case I am looking at, but I know that most mobile devices follow that access pattern as clients, and it would surprise me if the server does anything other than what the client asks.

Also the symptoms I am seeing match what I would expect given the (undocumented) ACD requests-per-hour throttling and the behaviour I am seeing when accessing higher bit rate streams. The reason why it manifests only on higher bit rate streams being that buffers are usually fixed size in MB, which means that there are fewer seconds per MB, thus more requests per hour on a higher bit rate stream. Since I can fetch an entire 50GB VM image down from in a single shot without any obvious slowdown, the throttling seems to be based on requests per hour rather than MB per hour.

But you are right, I'll put together some logging to confirm it beyond any doubt.

**Comment by gordan-bobic** [comment]
I believe this ticket is the same as the issue discussed in:
https://github.com/ncw/rclone/issues/1043

So these two issues should be merged.

**Comment by geoah** [comment]
@ncw have you given any thought as to how this would look like?

From my limited experience with `rclone` (couple of hours now) the best way to tackle this (and probably #1172 and #1043 at the same time) seems to be to create something similar to the existing `asyncReader` that instead of creating a pool of buffers for a single request it would create a pool of buffers per object. Buffers can be either in-memory or on-disk.

Unline `WithBuffer()` does now, but it would probably wrap `object.Open()` as well so it can check to see if there is an existing cache of the object in the pool, and if we have already cached the offset the user is seeking for, we return a reader for it.

The returned readers would probably also be used to tell the buffer to fetch more as it gets closer to where we have stopped fetching.

There are various issues to be solved, wrinkles to be ironed out etc here, but I'm just trying to get the ball rolling, figure out a design, and see if I can give this a shot. :)

Thanks.
